### PR TITLE
Feature: Allow users to tinker with RocksDB configuration via file; Upgrade RocksDb

### LINKED ISSUE #1624: Allow users to manually tweak RocksDb Options
### Description
Inside `mainnetdb` folder there is an Options file. It has all rocksdb configurations. Currently every time one restarts IRI, its contents is being replaced by hard coded options.
We want to ignore the hardcoded options in case the Options file is present.

### Motivation
One would like to tune RocksDb performance without changing java code.
Anyone in the community could fine settings that are good for him.

### Requirements
1. If options file is missing use hardcoded options
2. Else use options in file

### LINKED ISSUE #1629: Use the latest version of RocksDB
### Description
Update to the latest version of RocksDB.

### Motivation
Bug fixes and improvements in the latest version of rocksDB. 

### Requirements
- [ ] We have the latest version of RocksDB in IRI.

### LINKED ISSUE #1630: Tinker with the Rocks DB cache settings and benchmark the I/O bottlenecks
### Description
Tinker with the Rocks DB cache settings and benchmark the I/O bottlenecks

### Motivation
Figure out whether we can utilize the current cache and make it more performant instead of writing our own cache for resolving I/O bottlenecks.

### Requirements
- [x] Make the RocksDB settings configurable.
- [x] Use the latest RocksDB version.
- [ ] Profile the RocksDB instance to determine the bottleneck.  *SProf should be used for this.*
- [ ] Change up the RocksDB cache settings. 
- [ ] Don't run it on your machine, but use on a Hetzner
- [ ] Post a result with the machine name you used for future reproduction/comparison

The test 
*You can use our own RocksDBBenchmark.java for this.* (if you find a way to lower its runtime you get a beer) - there's a 1by1 fetch and reading in batch test
- Compare with the existing RocksDB values that IRI is enforcing
- Compare with the changes that you make to the cache - increasing its size, changing other properties that you find relevant.

### PR DESCRIPTION
**Pull request by kwek20**
# Description

In order to easily change rocksdb settings, you can now configure it directly though a config file.
Current settings but in a config would look like this:
```
create_if_missing=true
create_missing_column_families=true
max_log_file_size=1048576
max_manifest_file_size=1048576
max_open_files=10000
max_background_compactions=1
allow_concurrent_memtable_write=true
max_subcompactions=[your_number_of_cpus]
```

Fixes #1630
Fixes #1629 
Fixes #1624

## Type of change
- Enhancement (a non-breaking change which adds functionality)

# How Has This Been Tested?
It loads the config

# Checklist:
- [X] My code follows the style guidelines for this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by kwek20** [comment]
Splitting up the methods indicated by codacy will in my opinion make it less readable.
Therefor i chose to ignore the suggestion, and this PR is ready for merge!

File: src/main/java/com/iota/iri/storage/rocksDB/RocksDBPersistenceProvider.java [hunk 1]
```diff
@@ -436,52 +478,29 @@ private void initDB(String path, String logPath, Map<String, Class<? extends Per
...
-                .setMaxBackgroundCompactions(1);
-
-            options.setMaxSubcompactions(Runtime.getRuntime().availableProcessors());
+            options = createOptions(logPath, configFile);
 
             bloomFilter = new BloomFilter(BLOOM_FILTER_BITS_PER_KEY);
-
-            BlockBasedTableConfig blockBasedTableConfig = new BlockBasedTableConfig().setFilter(bloomFilter);
+            cache = new LRUCache(cacheSize * SizeUnit.KB, 2);
+            compressedCache = new LRUCache(32 * SizeUnit.KB, 10);
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/storage/rocksDB/RocksDBPersistenceProvider.java]: Those changes resulted in performance benefits? Or was it something you had to do due to RocksDb upgrade

File: src/main/java/com/iota/iri/storage/rocksDB/RocksDBPersistenceProvider.java [hunk 2]
```diff
@@ -70,22 +80,51 @@
...
+            Map.Entry<String, Class<? extends Persistable>> metadataColumnFamily) {
+        this(dbPath, logPath, null, cacheSize, columnFamilies, metadataColumnFamily);
+    }
+    
+    /**
+     * Creates a new RocksDB provider by reading the configuration to be used in this instance from a file
+     * 
+     * @param dbPath The location where the database will be stored
+     * @param logPath The location where the log files will be stored
+     * @param configFile The location where the RocksDB config is read from
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/storage/rocksDB/RocksDBPersistenceProvider.java]:
```suggestion
     * @param configPath The location where the RocksDB config is read from
```

File: src/main/java/com/iota/iri/storage/rocksDB/RocksDBPersistenceProvider.java [hunk 3]
```diff
@@ -530,4 +548,56 @@ private void initClassTreeMap(List<ColumnFamilyDescriptor> columnFamilyDescripto
...
+                Properties configProperties = new Properties();
+                InputStream stream = new FileInputStream(config);
+                try {
+                    configProperties.load(stream);
+                    options = DBOptions.getDBOptionsFromProps(configProperties);
+                } catch (IllegalArgumentException e) {
+                    log.warn("RocksDB configuration file is empty, falling back to default values");
+                } finally {
+                    stream.close();
+                }
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/storage/rocksDB/RocksDBPersistenceProvider.java]: I always like to use "try with resources" e.g.
`try (InputStream stream = new FileInputStream(config))`

Just a bit less verbose. What you did works as well, but I'd rather we use the shorter way throughout the project.

**Comment by GalRogozinski** [comment]
Codacy is still complaining but this is due to existing code you copy pasted, so it is fine

File: src/main/java/com/iota/iri/storage/rocksDB/RocksDBPersistenceProvider.java [hunk 4]
```diff
@@ -530,4 +548,54 @@ private void initClassTreeMap(List<ColumnFamilyDescriptor> columnFamilyDescripto
...
+                .setMaxSubcompactions(Runtime.getRuntime().availableProcessors());
+        }
+        
+        if (!BaseIotaConfig.Defaults.DB_LOG_PATH.equals(logPath) && logPath != null) {
+            if (!options.dbLogDir().equals("")) {
+                log.warn("Defined a db log path in config and commandline; Using the command line setting."); 
+            }
+            
+            options.setDbLogDir(logPath);
+        }
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/storage/rocksDB/RocksDBPersistenceProvider.java]: This is a bit late, I just noticed this code while merging conflicts...
Why is this here @kwek20 ?

Comment (kwek20) [reply src/main/java/com/iota/iri/storage/rocksDB/RocksDBPersistenceProvider.java]: This is here due to the fact we can define the db log on the config file and the logbath on iri.ini/cmdline. In order to warwn the user they defined it twice, we add a log
